cloud/WFLY-11824_wildfly_operator_kubernetes.adoc

```diff
@@ -0,0 +1,126 @@
+= WildFly Operator for Kubernetes/OpenShift
+:author:            Jeff Mesnil
+:email:             jmesnil@redhat.com
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Managing an application running on WildFly in a cloud environment can be a complex task. Things like ensuring containers have access to necessary resources, scaling up and scale down efficiently, integrating with outside services and handling rollout of new versions are all complex tasks that require operation knowledge. An https://coreos.com/blog/introducing-operators.html[Operator] is a specialized form of Kubernetes / OpenShift controller that is meant to encode that kind of operation knowledge in software, allowing users to more easily manage their cloud applications. The goal of this feature is to develop and Operator for managing WildFly-based applications.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-11824[WFLY-11824]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-1149[EAP7-1149]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+ * TBD
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+[ ] QE
+
+=== Affected Projects or Components
+
+ * https://github.com/wildfly/wildfly[WildFly]
+ * https://github.com/openshift-s2i/s2i-wildfly[WildFly s2i]
+
+
+=== Other Interested Projects
+
+ * https://github.com/jboss-dockerfiles/wildfly[WildFly Docker image]
+ * Keycloak / Red Hat SSO
+ * RHPAM
+
+== Requirements
+
+
+=== Hard Requirements
+
+==== Create a Custom Resource Definition (CRD) for WildFly.
+
+Provide a Custom Resource Definition (CRD) for an application based on a WildFly application server.
+
+The CRD will allow the user to:
+
+* Specify the name an application image that incorporates a WildFly server and an deployment that will be deployed by that server.
+** The name can include either a fixed or floating tag.
+* Specify an optional ConfigMap where the WildFly standalone configuration can be read from.
+* Specify the desired number of replicas of that image
+* Specify whether the replicas are intended to form a High Availability cluster (default is true).
+* Specify whether the containers are intended to be 'stateful'
+** default is to be determined (might be inferred from the server configuration)
+* Specify information for interfaces for which a Service and Route should be established
+** perhaps expose 8080 by default with an option to disable that
+** same for 9990 to expose management interface (including health check and metrics endpoints)
```

Comment:
Should management be accessible from outside the cluster? IMHO the wildfly operator should contribute a HAL console to the OpenShift as an extension and to a kubernetes cluster as a standalone management interface, so that users can manage the cluster in an easy/centralized/aggregated way. HAL should be the one providing proxied/aggregated APIs from the outside world? Why, because this way you can provide aggregated information from wildfly clusters, and not having to aggregate information on the client side, if every instance exposes its own management endpoint.

Reply:
@hpehl fiy. Depending on how HAL can be integrated with OpenShift, the operator would have to be enhanced/configured to take HAL into account.

Reply:
I think the operator should have such an option. The way how HAL is integrated into OpenShift is still wip, but the operator should definitely make it possible to include HAL and make it accessible from outside the cluster.